examples-next/auth/schema.ts

```diff
@@ -0,0 +1,144 @@
+ * control arguments; we need to revisit that in the context of how graphql schema extensions work
+ * now. See https://www.keystonejs.com/api/access-control#custom-schema-access-control
+ *
+ * Also note, I'm not sure the reasoning behind why ListAccess seems to be resolved once per
+ * mutation for mutations that operate on multiple items, while FieldAccess seems to be resolved
+ * once per item (many times per mutation) for the same. At least, that's what I think is going on.
+ *
+ * This seems like a real foot-gun, and I'd prefer that they're implemented consistently, in favor
+ * of aligning on FieldAccessInput and resolving the function once per item being updated in the
+ * case of mutations that operate on multiple items.
+ *
+ * Or maybe my understanding is wrong? This is what's implied by the documentation (see
+ * https://www.keystonejs.com/api/access-control#field-level-access-control) but then if it's not
+ * the case I don't see how `existingItem` works in mutations that operate on multiple items.
+ *
+ * I also have questions about `originalInput` in the case of nested mutations (e.g creating an
+ * item as the input to a relationship field). Is it the whole "original" input? Or just the nested
+ * input for the relationship? Seems like it would have to be the second, but it's not clear from
+ * the documentation. Is this consistent with how the `gqlName` argument works?
+ *
+ * tbh I'm inclined to remove the gqlName argument unless there's a compelling reason to have it;
+ * esp. given it's marked as optional.
+ *
+ * Finally, what's the point of the `auth` operation? It seems to be related to the listAuth
+ * functionality that we've just replaced with the new auth & session funcationality, if so we can
+ * remove it from access control in the new interfaces.
+ *
+ * Based on the docs (see https://www.keystonejs.com/api/access-control#the-auth-operation) it
+ * doesn't look like we need it anymore.
+ */
+
+type GenericItem = { id: string | number; [path: string]: any };
+type Session = any; // we have this defined elsewhere
+type Input = Record<string, any>;
+
+type ListAccessInput = {
+  session?: Session;
+  listKey: string;
+} & (
+  | { operation: 'create'; input: Input }
+  | { operation: 'read'; item: GenericItem }
+  | { operation: 'update'; item: GenericItem; input: Input }
+  | { operation: 'delete'; item: GenericItem }
+);
+
+type FieldAccessInput = {
+  session?: Session;
+  listKey: string;
+  fieldPath: string;
+} & (
+  | { operation: 'create'; input: Input }
+  | { operation: 'read'; item: GenericItem }
+  | { operation: 'update'; item: GenericItem; input: Input }
+);
+
+export const lists = createSchema({
+  User: list({
+    access: {
+      // Only allow admins to delete users
+      delete: ({ session }) => session?.data?.isAdmin,
+    },
+    admin: {
+      // Since you can't delete users unless you're an admin, we hide the UI for it
+      hideDelete: ({ session }) => session?.data?.isAdmin,
```

Comment:
Wouldn't this hide the delete for admins, but show delete for non-admins? I think that's the inverse of the intention.

Reply:
Good catch, thanks 😀